fapolicy_analyzer/tests/test_fapd_dbase.py

```diff
@@ -1,13 +1,98 @@
-from subprocess import SubprocessError
-from util.fapd_dbase import fapd_dbase_snapshot
+from subprocess import SubprocessError, Popen
+import pytest
+import os
+import glob
+import logging
+from pathlib import Path
+from fapolicy_analyzer.util.fapd_dbase import (
+    fapd_dbase_snapshot,
+    fapd_dbase_cleanup_snapshots,
+    gstrBackupBasename,
+    giBackupFileMaxCount,
+)
+from fapolicy_analyzer.util.xdg_utils import xdg_data_dir_prefix, xdg_config_dir_prefix
 
 
-def test_fapd_dbase_snapshot_fails():
+def delete_fapd_archives():
+    # Remove all fapd archive files w/default name
+    strArchivePathBasename = xdg_data_dir_prefix(gstrBackupBasename)
+    strSearchPattern = strArchivePathBasename + "_*.tgz"
+    listBackupFiles = glob.glob(strSearchPattern)
+    logging.debug(f"listBackupFiles = {listBackupFiles}")
+    if listBackupFiles:
+        # Iterate through the list and delete each file to empty dir.
+        for f in listBackupFiles:
+            if os.path.isfile(f):
+                # To address 'rm -i' in some envs where os.remove() will hang
+                listRmCmd = ["/bin/rm", "-f", f]
+                p = Popen(listRmCmd)
+                p.wait()
+
+
+def count_fapd_archives():
+    strArchivePathBasename = xdg_data_dir_prefix(gstrBackupBasename)
+    strSearchPattern = strArchivePathBasename + "_*.tgz"
+    listBackupFiles = glob.glob(strSearchPattern)
+    logging.debug(f"listBackupFiles = {listBackupFiles}")
+    return len(listBackupFiles)
+
+
+def create_populate_manifest():
+    strListFile = xdg_config_dir_prefix("fapolicy_backup_manifest.txt")
+    strDir = os.path.dirname(strListFile)
+    strHome = os.environ.get("HOME")
+    Path(strDir).mkdir(parents=True, exist_ok=True)
+
+    with open(strListFile, "w") as fileTarInputList:
+        strArchiveDirs = strHome + "/.config/\n" + strHome + "/.cargo/\n"
+        fileTarInputList.write(strArchiveDirs)
+    return strListFile
+
+
+def delete_manifest():
+    pass
+    # strListFile = xdg_config_dir_prefix("fapolicy_backup_manifest.txt")
+    # os.remove(strListFile)
+
+
+@pytest.fixture
+def manifest():
+    strListFile = create_populate_manifest()
+    yield strListFile
+    delete_manifest()
+
+
+def test_fapd_dbase_snapshot_and_cleanup(manifest):
+    # Mock the subprocess.Popen() call to throw an exception
+    delete_fapd_archives()
+
+    # Verify no fapd archives
+    assert count_fapd_archives() == 0
+
+    # Create and verify one fapd archive
+    assert fapd_dbase_snapshot()
+    assert count_fapd_archives() == 1
+
+    # Add another giBackupFileMaxCount+1 of files. Verify max count limit
+    for i in range(giBackupFileMaxCount + 1):
+        assert fapd_dbase_snapshot()
+    assert count_fapd_archives() == giBackupFileMaxCount
+
+    # Clean up: Delete and verify all archives have been removed.
+    delete_fapd_archives()
+    assert count_fapd_archives() == 0
```

Comment:
It seems like it would be easier to mock the `subprocess.Popen` here and then assert on the calls to it instead of writing all these helper methods to interact with the file system.